mosaic/mosaic.go

```diff
@@ -0,0 +1,602 @@
+		{Char: '▝', Coverage: [4]bool{false, true, false, false}, CoverageMap: " █\n  "}, // Quadrant upper right.
+		{Char: '▖', Coverage: [4]bool{false, false, true, false}, CoverageMap: "  \n█ "}, // Quadrant lower left.
+		{Char: '▗', Coverage: [4]bool{false, false, false, true}, CoverageMap: "  \n █"}, // Quadrant lower right.
+		{Char: '▌', Coverage: [4]bool{true, false, true, false}, CoverageMap: "█ \n█ "},  // Left half block.
+		{Char: '▐', Coverage: [4]bool{false, true, false, true}, CoverageMap: " █\n █"},  // Right half block.
+		{Char: '▀', Coverage: [4]bool{true, true, false, false}, CoverageMap: "██\n  "},  // Upper half block (already added).
+		{Char: '▄', Coverage: [4]bool{false, false, true, true}, CoverageMap: "  \n██"},  // Lower half block (already added).
+	}
+	ComplexBlocks = []Block{
+		{Char: '▙', Coverage: [4]bool{true, false, true, true}, CoverageMap: "█ \n██"},  // Quadrant upper left and lower half.
+		{Char: '▟', Coverage: [4]bool{false, true, true, true}, CoverageMap: " █\n██"},  // Quadrant upper right and lower half.
+		{Char: '▛', Coverage: [4]bool{true, true, true, false}, CoverageMap: "██\n█ "},  // Quadrant upper half and lower left.
+		{Char: '▜', Coverage: [4]bool{true, true, false, true}, CoverageMap: "██\n █"},  // Quadrant upper half and lower right.
+		{Char: '▚', Coverage: [4]bool{true, false, false, true}, CoverageMap: "█ \n █"}, // Quadrant upper left and lower right.
+		{Char: '▞', Coverage: [4]bool{false, true, true, false}, CoverageMap: " █\n█ "}, // Quadrant upper right and lower left.
+	}
+)
+
+// Block represents different Unicode block characters.
+type Block struct {
+	Char        rune
+	Coverage    [4]bool // Which parts of the block are filled (true = filled).
+	CoverageMap string  // Visual representation of coverage for debugging.
+}
+
+// Symbol represents the symbol type to use when rendering the image.
+type Symbol string
+
+// Symbol types.
+const (
+	AllSymbols     Symbol = "all"
+	HalfSymbols    Symbol = "half"
+	QuarterSymbols Symbol = "quarter"
+)
+
+// Scale represents scale mode that will be used when rendering the image.
+type Scale string
+
+// Symbol types.
+const (
+	Fit     Scale = "all"
+	None    Scale = "none"
+	Stretch Scale = "stretch"
+	Center  Scale = "center"
+)
+
+// Options contains all configurable settings.
+type Mosaic struct {
+	Blocks         []Block
+	colorMode      int     // 0=none, 1=8colors, 2=256colors, 3=truecolor.
+	OutputWidth    int     // Output width.
+	OutputHeight   int     // Output height (0 for auto).
+	ThresholdLevel uint8   // Threshold for considering a pixel as set (0-255).
+	DitherLevel    float64 // Dithering amount (0.0-1.0).
+	UseFgBgOnly    bool    // Use only foreground/background colors (no block symbols).
+	InvertColors   bool    // Invert colors.
+	ScaleMode      Scale   // fit, stretch, center.
+	Symbols        Symbol  // Which symbols to use: "half", "quarter", "all".
+}
+
+// Create Mosaic
+func New() *Mosaic {
```

Comment:
We don't want to use pointers so that copying becomes easier.
```go
mosiacA := mosiac.New().Scale(None)
mosiacB := mosiacA.Scale(Fit) // creates a copy without modifying mosaicA
```